File: arrays_and_strings/1.1.rb
# Implement an algorithm to determine if a string has all unique characters. What if you cannot use additional data structures?

# for each character in the string
#   if any subsequent character matches it
#     fail
# succeed

# Algorithm 1
# Use array's unique method and compare size to original string size
def has_dup?(s)
  s = s.split('')
  s.uniq.size != s.size
  # .uniq returns a new array by removing duplicate values in self.
end

# Algorithm 2
# Sort characters first, then check if next character is equal to check uniqueness
def sort_check_unique(s)
  s = s.chars.sort.join
  (0...s.size).each do |i|
    return false if s[i] == s[i+1]
  end
  true
end

File: arraysStrings/1.1.js
// Is Unique:

// Implement an algorithm to determine if a string has all unique characters.

function isUnique(str){
  var element = {};
  for (var i = 0; i < str.length; i++ ){
    if (element[str[i]]){
      return false;
    }
    element[str[i]] = true;
  }
  return true;
}

// OR 

function isUnique(str){
  var hash = {};
  for (var i = 0; i < str.length; i++ ){
    var ele = str[i];
    if(!hash[ele]){
     hash[ele] = 1;
   } else {
     hash[ele]++;
    }
  }
  for(var key in hash){
    if(hash[key] > 1){
      return false;
    }
  }
  return true;
}
